### PR TITLE
Hbsun/end2end demo

### PR DESCRIPTION
**Pull request by cuhk-hbsun**

### CODE REVIEW

File: demo/end2end_image_demo.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        end2end_res['result'].append(box_res)
+
+    return end2end_res
+
+
+def main():
+    parser = ArgumentParser()
+    parser.add_argument('img', type=str, help='Input Image file.')
+    parser.add_argument(
+        'save_path', type=str, help='Path to save visualized image.')
```
Comment (jeffreykuang) [inline demo/end2end_image_demo.py:44]:
```suggestion
        'out-file', type=str, help='Output file name of the visualized image.')
```

File: demo/end2end_image_demo.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    parser.add_argument(
+        '--detect-model',
+        type=str,
+        default='psenet',
+        choices=['psenet', 'panet', 'dbnet', 'textsnake', 'maskrcnn'],
+        help='Text detect model type.')
+    parser.add_argument(
+        '--recog-model',
+        type=str,
+        default='sar',
```
Comment (jeffreykuang) [inline demo/end2end_image_demo.py:54]:
```suggestion
        default='crnn',
```

File: demo/end2end_image_demo.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    parser.add_argument('img', type=str, help='Input Image file.')
+    parser.add_argument(
+        'out_file', type=str, help='Output file name of the visualized image.')
+    parser.add_argument(
+        '--detect-model',
+        type=str,
+        default='psenet',
+        choices=['psenet', 'panet', 'dbnet', 'textsnake', 'maskrcnn'],
+        help='Text detect model type.')
+    parser.add_argument(
```
Comment (jeffreykuang) [inline demo/end2end_image_demo.py:55]: I suggest to add 5 more argument so that user can specify any config and model they want to test.

--detect-config
--detect-ckpt
--recog-config
--recog-ckpt
--recog-rgb (whether use rgb in reognition model)

if the above arguments is set, then we use the model user specified. Otherwise, we use --recog-alg
and --detect-alg

File: demo/end2end_image_demo.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        '--recog-model',
```
Comment (jeffreykuang) [inline demo/end2end_image_demo.py:56]:
```suggestion
        '--recog-alg',
```

File: demo/end2end_image_demo.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    return end2end_res
+
+
+def main():
+    parser = ArgumentParser()
+    parser.add_argument('img', type=str, help='Input Image file.')
+    parser.add_argument(
+        'out_file', type=str, help='Output file name of the visualized image.')
+    parser.add_argument(
+        '--detect-model',
```
Comment (jeffreykuang) [inline demo/end2end_image_demo.py:50]:
```suggestion
        '--detect-alg',
```

File: mmocr/core/end2end_visualize.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            thickness=cv2.FILLED)
+        out_img = cv2.addWeighted(dst_img, 0.5, img, 0.5, 0)
+    return out_img
+
+
+def get_optimal_font_scale(text, width):
+    """Get optimal font scale for cv2.putText.
+
+    Args:
+        text (str): Text in box.
```
Comment (jeffreykuang) [inline mmocr/core/end2end_visualize.py:51]:
```suggestion
        text (str): Text in one box.
```

File: mmocr/core/end2end_visualize.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        width (int): Width of box.
```
Comment (jeffreykuang) [inline mmocr/core/end2end_visualize.py:52]:
```suggestion
        width (int): The box width.
```

File: mmocr/datasets/pipelines/crop.py [hunk 8]
```diff
@@ -83,24 +83,39 @@ def warp_img(src_img,
     return dst_img
 
 
-def crop_img(src_img, box):
+def crop_img(src_img, box, long_edge_pad_ratio=0.4, short_edge_pad_ratio=0.2):
     """Crop box area to rectangle.
```
Comment (jeffreykuang) [inline mmocr/datasets/pipelines/crop.py:87]:
```suggestion
    """Crop text regions with their bounding boxes.
```

File: mmocr/datasets/pipelines/crop.py [hunk 9]
```diff
@@ -83,24 +83,39 @@ def warp_img(src_img,
...
 
     Args:
         src_img (np.array): Image before crop.
```
Comment (jeffreykuang) [inline mmocr/datasets/pipelines/crop.py:90]:
```suggestion
        src_img (np.array): The original image.
```

File: demo/end2end_image_demo.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
+from argparse import ArgumentParser
+
+import mmcv
+
+from mmdet.apis import init_detector
+from mmocr.apis.inference import model_inference
+from mmocr.core.end2end_visualize import end2end_show_result, write_json
+from mmocr.core.prepare_model import prepare_det_model, prepare_recog_model
+from mmocr.datasets.pipelines.crop import crop_img
+from mmocr.models import build_detector  # noqa: F401
```
Comment (innerlee) [inline demo/end2end_image_demo.py:10]: Is it possible to remove # noqa: F401

File: demo/end2end_image_demo.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
+from argparse import ArgumentParser
```
Comment (innerlee) [inline demo/end2end_image_demo.py:1]: The filename could be better :)

File: demo/end2end_image_demo.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+
+from mmdet.apis import init_detector
+from mmocr.apis.inference import model_inference
+from mmocr.core.end2end_visualize import end2end_show_result, write_json
+from mmocr.core.prepare_model import prepare_det_model, prepare_recog_model
+from mmocr.datasets.pipelines.crop import crop_img
+from mmocr.models import build_detector  # noqa: F401
+
+
+def end2end_inference(args, det_model, recog_model):
```
Comment (innerlee) [inline demo/end2end_image_demo.py:13]: `det_and_recog_inference` or `inference_det_and_recog`?

File: demo/end2end_image_demo.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+    end2end_res['result'] = []
+
+    image = mmcv.imread(image_path)
+    det_result = model_inference(det_model, image)
+    bboxes = det_result['boundary_result']
+
+    for bbox in bboxes:
+        box_res = {}
+        box_res['box'] = [round(x) for x in bbox[:-1]]
+        box_res['box_score'] = float('{:.4f}'.format(bbox[-1]))
```
Comment (innerlee) [inline demo/end2end_image_demo.py:24]: What's the purpose of `float('{:.4f}'.format(bbox[-1]))`? round?

File: demo/end2end_image_demo.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        box = bbox[:8]
+        if len(bbox) > 9:
+            min_x = min(bbox[0:-1:2])
+            min_y = min(bbox[1:-1:2])
+            max_x = max(bbox[0:-1:2])
+            max_y = max(bbox[1:-1:2])
+            box = [min_x, min_y, max_x, min_y, max_x, max_y, min_x, max_y]
+        box_img = crop_img(image, box)
+
+        if args.recog_alg == 'crnn':
```
Comment (innerlee) [inline demo/end2end_image_demo.py:34]: Shouldn't these conversion be done via pipelines?

File: demo/end2end_image_demo.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+            box_img = mmcv.bgr2gray(box_img, keepdim=True)
+
+        recog_result = model_inference(recog_model, box_img)
+
+        text = recog_result['text']
+        text_score = recog_result['score']
+        if isinstance(text_score, list):
+            text_score = sum(text_score) / max(1, len(text))
+        box_res['text'] = text
+        box_res['text_score'] = float('{:.4f}'.format(text_score))
```
Comment (innerlee) [inline demo/end2end_image_demo.py:44]: and here

File: demo/end2end_image_demo.py (hunk 3 above)
Comment (innerlee) [reply demo/end2end_image_demo.py:61]: Yeah pass configs instead of algorithm names

File: mmocr/core/prepare_model.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
+import os
+import os.path as osp
+import shutil
+import urllib
+
+
+def prepare_ckpt(checkpoint_file, checkpoint_url):
```
Comment (innerlee) [inline mmocr/core/prepare_model.py:7]: These shouldn't exist.
Pls refactor the code as jeffery commented https://github.com/open-mmlab/mmocr/pull/105/files#r617689124

File: tests/test_models/test_ocr_backbone.py [hunk 17]
```diff
@@ -1,7 +1,8 @@
 import pytest
 import torch
 
-from mmocr.models.textrecog.backbones import ResNet31OCR, VeryDeepVgg
+from mmocr.models.textrecog.backbones import (NRTRModalityTransform,
```
Comment (innerlee) [inline tests/test_models/test_ocr_backbone.py:4]: Unrelated change? How about putting them in a separate pr?

**Comment by yauauau** [comment]
Ok stop sand

File: mmocr/datasets/pipelines/crop.py [hunk 18]
```diff
@@ -83,24 +83,39 @@ def warp_img(src_img,
...
+        long_edge_pad_ratio (float): Box pad ratio for long edge.
+        short_edge_pad_ratio (float): Box pad ratio for short edge.
     """
     assert utils.is_type_list(box, float) or utils.is_type_list(box, int)
     assert len(box) == 8
+    assert 0. <= long_edge_pad_ratio < 1.0
+    assert 0. <= short_edge_pad_ratio < 1.0
 
     h, w = src_img.shape[:2]
     points_x = [min(max(x, 0), w) for x in box[0:8:2]]
```
Comment (innerlee) [inline mmocr/datasets/pipelines/crop.py:101]: np.clip

File: mmocr/datasets/pipelines/crop.py [hunk 19]
```diff
@@ -83,24 +83,39 @@ def warp_img(src_img,
...
     points_y = [min(max(y, 0), h) for y in box[1:9:2]]
 
-    left = int(min(points_x))
-    top = int(min(points_y))
-    right = int(max(points_x))
-    bottom = int(max(points_y))
+    box_width = max(points_x) - min(points_x)
+    box_height = max(points_y) - min(points_y)
+    if box_height < box_width:
```
Comment (innerlee) [inline mmocr/datasets/pipelines/crop.py:106]: font_size = min(box_width, box_height)

File: mmocr/datasets/pipelines/crop.py [hunk 20]
```diff
@@ -83,24 +83,39 @@ def warp_img(src_img,
...
-def crop_img(src_img, box):
-    """Crop box area to rectangle.
+def crop_img(src_img, box, long_edge_pad_ratio=0.4, short_edge_pad_ratio=0.2):
+    """Crop text region with their bounding box.
 
     Args:
-        src_img (np.array): Image before crop.
+        src_img (np.array): The original image.
         box (list[float | int]): Points of quadrangle.
+        long_edge_pad_ratio (float): Box pad ratio for long edge.
```
Comment (innerlee) [inline mmocr/datasets/pipelines/crop.py:92]: Explain when will padding happen, and explain the size that "ratio" depend on

File: mmocr/datasets/pipelines/crop.py [hunk 21]
```diff
@@ -83,24 +83,39 @@ def warp_img(src_img,
...
+    if box_height < box_width:
+        font_size = box_height
+        horizontal_pad = long_edge_pad_ratio * font_size
+        vertical_pad = short_edge_pad_ratio * font_size
+    else:
+        font_size = box_width
+        horizontal_pad = short_edge_pad_ratio * font_size
+        vertical_pad = long_edge_pad_ratio * font_size
+
+    left = int(min(points_x) - horizontal_pad)
```
Comment (innerlee) [inline mmocr/datasets/pipelines/crop.py:115]: does left need to >=0?

**Comment by innerlee** [comment]
Hi @yauauau , it seems that you have watched the repo. Please unwatch it to get less emails.

File: docs/getting_started.md [hunk 22]
```diff
@@ -24,6 +24,14 @@ python demo/image_demo.py demo/demo_text_det.jpg configs/xxx.py xxx.pth demo/dem
 
 The predicted result will be saved as `demo/demo_text_det_pred.jpg`.
 
+To end-to-end test a single image with both text detection and recognition,
+
+```shell
+python demo/det_recog_image_demo.py demo/demo_text_det.jpg demo/output.jpg
```
Comment (innerlee) [inline docs/getting_started.md:30]: How about "ocr_image_demo.py"

File: demo/ocr_image_demo.py [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+    return end2end_res
+
+
+def main():
+    parser = ArgumentParser()
+    parser.add_argument('img', type=str, help='Input Image file.')
+    parser.add_argument(
+        'out_file', type=str, help='Output file name of the visualized image.')
+    parser.add_argument(
+        '--detect-config',
```
Comment (innerlee) [inline demo/ocr_image_demo.py:52]: det_config or detection_config

File: mmocr/core/det_recog_visualize.py [hunk 24]
```diff
@@ -0,0 +1,121 @@
+import json
```
Comment (innerlee) [inline mmocr/core/det_recog_visualize.py:1]: visualization.py

File: mmocr/datasets/pipelines/loading.py [hunk 25]
```diff
@@ -66,3 +67,39 @@ def _load_masks(self, results):
...
+        Args:
+            results (dict): Result dict with Webcam read image in
+                ``results['img']``.
+
+        Returns:
+            dict: The dict contains loaded image and meta information.
+        """
+
+        img = results['img']
+        if self.to_float32:
```
Comment (innerlee) [inline mmocr/datasets/pipelines/loading.py:92]: mv this after L97, since float32 will affect how bgr2gray or gray2bgr works.

For simplicity, you can assert that results['img'] is uint8, since otherwise, things might go wrong with color conversions

File: mmocr/datasets/pipelines/loading.py [hunk 26]
```diff
@@ -66,3 +67,39 @@ def _load_masks(self, results):
         results['gt_masks'] = gt_masks
         results['mask_fields'].append('gt_masks')
         return results
+
+
+@PIPELINES.register_module()
+class LoadImageFromMat(LoadImageFromFile):
+    """Load an image from cv Mat.
```
Comment (innerlee) [inline mmocr/datasets/pipelines/loading.py:74]: I don't think we have Mat type. They are numpy arrays

File: mmocr/datasets/pipelines/loading.py [hunk 27]
```diff
@@ -66,3 +67,39 @@ def _load_masks(self, results):
...
+
+        img = results['img']
+        if self.to_float32:
+            img = img.astype(np.float32)
+        if self.color_type == 'grayscale' and img.shape[2] == 3:
+            img = mmcv.bgr2gray(img, keepdim=True)
+        if self.color_type == 'color' and img.shape[2] == 1:
+            img = mmcv.gray2bgr(img)
+
+        results['filename'] = None
```
Comment (innerlee) [inline mmocr/datasets/pipelines/loading.py:99]: Why drops the 'filename' info (if exists)

File: mmocr/core/__init__.py [hunk 28]
```diff
@@ -1,3 +1,4 @@
+from .det_recog_visualize import *  # noqa: F401, F403
```
Comment (innerlee) [inline mmocr/core/__init__.py:1]: Since you `import *` from  det_recog_visualize, the internal functions should be prefix-ed with `_`.

File: mmocr/core/det_recog_visualize.py [hunk 29]
```diff
@@ -0,0 +1,121 @@
...
+
+import cv2
+import mmcv
+import numpy as np
+
+
+def write_json(obj, fpath):
```
Comment (innerlee) [inline mmocr/core/det_recog_visualize.py:8]: This function do not belong to visualization

Comment (innerlee) [reply mmocr/core/det_recog_visualize.py:8]: mmcv can dump json to files btw

File: mmocr/datasets/pipelines/loading.py (hunk 27 above)
Comment (cuhk-hbsun) [reply mmocr/datasets/pipelines/loading.py:99]: Just a cropped image of numpy array, no filename. This func is modified from `LoadImageFromWebcam` in mmdetection, which not support color conversion (maybe a bug)

Comment (innerlee) [reply mmocr/datasets/pipelines/loading.py:100]: Okay fine. The name contains "Load" so this is reasonable

**Review by innerlee — APPROVED** [review]
